go/vt/vttablet/onlineddl/executor.go

```diff
@@ -1185,30 +1454,176 @@ func (e *Executor) dropPTOSCMigrationTriggers(ctx context.Context, onlineDDL *sc
 	return err
 }
 
-// reviewRunningMigrations iterates migrations in 'running' state (there really should just be one that is
-// actually running).
-func (e *Executor) reviewRunningMigrations(ctx context.Context) (countRunnning int, runningNotByThisProcess []string, err error) {
-	e.migrationMutex.Lock()
-	defer e.migrationMutex.Unlock()
+// readVReplStream reads _vt.vreplication entries for given workflow
+func (e *Executor) readVReplStream(ctx context.Context, uuid string, okIfMissing bool) (*VReplStream, error) {
+	query, err := sqlparser.ParseAndBind(sqlReadVReplStream,
+		sqltypes.StringBindVariable(uuid),
+	)
+	if err != nil {
+		return nil, err
+	}
+	r, err := e.execQuery(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	if len(r.Rows) == 0 && okIfMissing {
+		return nil, nil
+	}
+	row := r.Named().Row()
+	if row == nil {
+		return nil, vterrors.Errorf(vtrpcpb.Code_UNKNOWN, "Cannot find unique workflow for UUID: %+v", uuid)
+	}
+	s := &VReplStream{
+		id:                   row.AsInt64("id", 0),
+		workflow:             row.AsString("workflow", ""),
+		source:               row.AsString("source", ""),
+		pos:                  row.AsString("pos", ""),
+		timeUpdated:          row.AsInt64("time_updated", 0),
+		transactionTimestamp: row.AsInt64("transaction_timestamp", 0),
+		state:                row.AsString("state", ""),
+		message:              row.AsString("message", ""),
+		bls:                  &binlogdatapb.BinlogSource{},
+	}
+	if err := proto.UnmarshalText(s.source, s.bls); err != nil {
+		return nil, err
+	}
+	return s, nil
+}
 
-	parsed := sqlparser.BuildParsedQuery(sqlSelectRunningMigrations, "_vt", ":strategy")
-	bindVars := map[string]*querypb.BindVariable{
-		"strategy": sqltypes.StringBindVariable(string(schema.DDLStrategyPTOSC)),
+// isVReplMigrationReadyToCutOver sees if the vreplication migration has completed the row copy
+// and is up to date with the binlogs.
+func (e *Executor) isVReplMigrationReadyToCutOver(ctx context.Context, s *VReplStream) (isReady bool, err error) {
+	// Check all the cases where migration is still running:
```

Comment:
When an online ddl completes, vttablets and vtgates will need to reload their schema as per the current mechanisms, before they see the changes made by the ddl correct?

Reply:
Right. I issue a `ReloadSchema` before running the migration. Should I run `ReloadSchema` immediately after completing the migration, or should I actually do that during the cut-over (when tables have been switched, and writes are still blocked)?

Reply:
Doing it when writes are blocked might be better since binlog events after the write will then see the new schema. Otherwise we may have a race: new events for the altered table may be processed before the new schema is reloaded.

But the `ReloadSchema` is just for this tablet, right?

Reply:
https://github.com/vitessio/vitess/pull/7419/files#diff-059c9f46e8d270d9c5514ef2b08679035eb0daaa8d95074e34ef43a81d50dc37R637-R639

correct

Reply:
Just clarifying, that this goes through gRPC; so I want to make sure this isn't an expensive event, because meanwhile we're blocking (rather -- rejecting) writes to the table, so it's a critical point in time.

Reply:
Define expensive...: ReloadSchema can take a while since it parses the entire db schema. It has been speeded up recently but for a large number of tables it will probably take a non-trivial amount of time. I would think, still in the order of milliseconds, but not entirely sure.

Reply:
> non-trivial amount of time.

milliseconds is just fine. Would it take more than 1 second (I'd define "expensive" at 1 second) ? e.g. on a schema with 10,000 tables? Is it possible to `ReloadSchema` for just one table?

Reply:
(having discussed this in sync with @rohit-nayak-ps) reloading schema can take a substantial time if many thousands of tables are involved. For now, we issue an async ReloadSchema (with goroutine). In the future, we will look into reloading a single table.